diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -28,7 +28,7 @@
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fdiagnostics-color=always")
 endif()
 
-find_package(Boost 1.70 CONFIG COMPONENTS context REQUIRED)
+find_package(Boost 1.70 COMPONENTS context REQUIRED)
 
 include(FetchContent)
 
diff --git a/cmake/simsycl-config.cmake.in b/cmake/simsycl-config.cmake.in
--- a/cmake/simsycl-config.cmake.in
+++ b/cmake/simsycl-config.cmake.in
@@ -14,7 +14,7 @@
 set(SIMSYCL_ORIGINAL_CMAKE_MODULE_PATH "${CMAKE_MODULE_PATH}")
 set(CMAKE_MODULE_PATH "${CMAKE_MODULE_PATH}" "${SIMSYCL_CMAKE_DIR}")
 
-find_dependency(Boost 1.70 CONFIG COMPONENTS context REQUIRED)
+find_dependency(Boost 1.70 COMPONENTS context REQUIRED)
 find_dependency(nlohmann_json)
 find_dependency(libenvpp)
 
